### PR TITLE
added ctime() code in time.java

### PR DESCRIPTION
**Pull request by prithvisingh18**
Implementing `ctime()` from `time.py`(standard library) in `time.java` .

### CODE REVIEW

File: python/common/python/time.java [hunk 1]
```diff
@@ -89,10 +89,26 @@ public time() {
     }
 
     @org.python.Method(
-            __doc__ = ""
+            __doc__ = "str(object='') -> str\n"+
```
Comment (eliasdorneles) [inline python/common/python/time.java]: This docstring seems completely unrelated to the one for the ``time.ctime`` function.

File: python/common/python/time.java [hunk 2]
```diff
@@ -89,10 +89,26 @@ public time() {
...
+                      "str(bytes_or_buffer[, encoding[, errors]]) -> str\n"+
+                      "Create a new string object from the given object. If encoding or"+
+                      "errors is specified, then the object must expose a data buffer"+
+                      "that will be decoded using the given encoding and error handler."+
+                      "Otherwise, returns the result of object.__str__() (if defined)"+
+                      "or repr(object)."+
+                      "encoding defaults to sys.getdefaultencoding().\n"+
+                      "errors defaults to 'strict'."
     )
     public static org.python.Object ctime() {
```
Comment (eliasdorneles) [inline python/common/python/time.java]: This function is supposed to take an optional argument: you can pass the number of seconds:

```
>>> import time
>>> time.ctime(0)
'Thu Jan  1 01:00:00 1970'
>>> now = time.time()
>>> time.ctime(now - now % 3600)
'Mon Feb 26 15:00:00 2018'
```

The above are good candidates for tests, btw.

File: python/common/python/time.java (hunk 1 above)
Comment (prithvisingh18) [reply python/common/python/time.java]: I agree but this is what python3 returns when `time.ctime().__doc__` .
Is there an alternative ?

File: python/common/python/time.java (hunk 2 above)
Comment (prithvisingh18) [reply python/common/python/time.java]: I have overloaded the function `ctime()` for optional parameter, what could be the reason for the following error, 
`TypeError: ctime() takes 0 positional arguments but 1 was given`
Thanks.

File: python/common/python/time.java (hunk 1 above)
Comment (eliasdorneles) [reply python/common/python/time.java]: That's giving you the docstring of the result of calling the function (which is a string).
Try `time.ctime.__doc__`.

File: python/common/python/time.java (hunk 2 above)
Comment (eliasdorneles) [reply python/common/python/time.java]: Check the docs section on Python signatures for Java-defined methods: http://voc.readthedocs.io/en/latest/reference/signatures.html

File: python/common/python/time.java [hunk 3]
```diff
@@ -106,7 +106,7 @@ public time() {
         return new org.python.types.Str(ft.format(currentDate));
     }
     public static org.python.Object ctime(org.python.Object seconds) {
-        java.util.Date currentDate = new java.util.Date(((org.python.types.Int) seconds.__int__()).value * 1000);
+        java.util.Date currentDate = new java.util.Date((long) (((org.python.types.Int) seconds.__int__()).value * 1000));
```
Comment (uranusjr) [inline python/common/python/time.java]: Would it be a good idea to cast to long *before* multiplying?

Comment (prithvisingh18) [reply python/common/python/time.java]: I am not sure if `int` would be large enough hence, I used `(long)` as the seconds could be a large.

Comment (uranusjr) [reply python/common/python/time.java]: I understand the line of thinking, but it’s the implementation I’m having doubts about. With the current implementation, the number is first times 1000, and then casted to long. This wouldn’t prevent anything because the cast happens after the overflow (if there is one).

To actually prevent the overflow you’d need to do something like

```java
new java.util.Date(((long) ((org.python.types.Int) seconds.__int__()).value) * 1000)
```

or (probably better)

```java
new java.util.Date(((org.python.types.Int) seconds.__int__()).value * 1000L)
```

so the large number is casted to long first, and then multiplied.

File: python/common/python/time.java [hunk 4]
```diff
@@ -89,10 +89,30 @@ public time() {
...
     )
-    public static org.python.Object ctime() {
-        throw new org.python.exceptions.NotImplementedError("time.ctime() has not been implemented.");
+    public static org.python.Object ctime(org.python.Object seconds) {
+        java.util.Date date;
+        if(seconds == null){
+            long currentTimeInMillis = System.currentTimeMillis();
+            date = new java.util.Date(currentTimeInMillis);
+        } else {
+            date = new java.util.Date((long) (((org.python.types.Int) seconds.__int__()).value * 1000));
```
Comment (eliasdorneles) [inline python/common/python/time.java]: Is the call to `__int__` really needed here?
Wouldn't simply `((org.python.types.Int) seconds).value` work already?

File: tests/stdlib/test_time.py [hunk 5]
```diff
@@ -106,12 +106,19 @@ def test_clock(self):
...
             import time
             print(time.ctime())
             """)
+    def test_ctime_with_parameter(self):
+        self.assertCodeExecution("""
+            import time
+            print(time.ctime(0))
+            print(time.ctime(1000))
+            now = time.time()
+            time.ctime((now - (now % 3600)))
```
Comment (eliasdorneles) [inline tests/stdlib/test_time.py]: This line is missing a print: if we don't print, the result isn't checked.

File: tests/stdlib/test_time.py [hunk 6]
```diff
@@ -106,12 +106,19 @@ def test_clock(self):
 
     #######################################################
     # ctime
-    @expectedFailure
     def test_ctime(self):
         self.assertCodeExecution("""
             import time
             print(time.ctime())
```
Comment (eliasdorneles) [inline tests/stdlib/test_time.py]: This test will be very hard to pass because it depends too much on the time of execution.

Your best bet for this test is to discard some of the precision, like: `print(time.ctime()[:10], time.ctime()[-4:])`

This is considering only the `date` information and discarding the hours. It's still a flaky test because it will fail when it runs at exactly midnight, but it's at least giving us good confidence that the code does what it's supposed to do.

Makes sense?

Comment (prithvisingh18) [reply tests/stdlib/test_time.py]: I've also had the same problem, seems right will do it now.

File: python/common/python/time.java (hunk 3 above)
Comment (prithvisingh18) [reply python/common/python/time.java]: Thanks.

**Comment by goanpeca** [comment]
ping ping @eliasdorneles  :-p

File: tests/stdlib/test_time.py [hunk 7]
```diff
@@ -106,12 +106,19 @@ def test_clock(self):
...
+            print(time.ctime()[:10], time.ctime()[-4:])
             """)
+    def test_ctime_with_parameter(self):
+        self.assertCodeExecution("""
+            import time
+            print(time.ctime(0))
+            print(time.ctime(1000))
+            now = time.time()
+            print(time.ctime((now - (now % 3600))))
+        """)
```
Comment (eliasdorneles) [inline tests/stdlib/test_time.py]: Can you add a test for the case of wrong parameter, like if we give it a string `time.ctime('today')` ?
It should show the same message as CPython (you'll need to do add a type-check for the argument).

File: python/common/python/time.java [hunk 8]
```diff
@@ -89,10 +89,30 @@ public time() {
...
+        java.text.SimpleDateFormat ft = new java.text.SimpleDateFormat("E MMM dd HH:mm:ss yyyy");
+        String padded_date = ft.format(date);
+        if(Character.toString(padded_date.charAt(8)).equals("0")) {
+            java.lang.StringBuilder sb = new StringBuilder();
+            sb.append(padded_date.substring(0,7));
+            sb.append("  ");
+            sb.append(padded_date.substring(9,padded_date.length()));
+            padded_date = sb.toString();
+        }
+        return new org.python.types.Str(padded_date);
```
Comment (eliasdorneles) [inline python/common/python/time.java]: Can you please run `ant checkstyle` locally and fix the style errors that it finds?
Normally we can see these in the CI, but [it seems to not be working correctly right now](https://github.com/pybee/voc/issues/640).

Comment (prithvisingh18) [reply python/common/python/time.java]: Yeah sure.

File: python/common/python/time.java [hunk 9]
```diff
@@ -89,10 +89,42 @@ public time() {
...
+            String parameterType = seconds.getClass().getSimpleName();
+            if (!parameterType.equals("Int") && !parameterType.equals("Float")) {
+                if (parameterType.equals("Str")) {
+                    parameterType = "str";
+                } else if (parameterType.equals("List")) {
+                    parameterType = "list";
+                } else if (parameterType.equals("Tuple")) {
+                    parameterType = "tuple";
+                }
+                throw new org.python.exceptions.TypeError("an integer is required (got type " + parameterType + ")");
```
Comment (eliasdorneles) [inline python/common/python/time.java]: As you can see, this will not work for all existing types.
There are several examples in the codebase on how to do this, look at the implementation of `List.insert` for example.

File: python/common/python/time.java [hunk 10]
```diff
@@ -89,10 +89,34 @@ public time() {
...
-    public static org.python.Object ctime() {
-        throw new org.python.exceptions.NotImplementedError("time.ctime() has not been implemented.");
+    public static org.python.Object ctime(org.python.Object seconds) {
+        java.util.Date date;
+        if (seconds == null) {
+            long currentTimeInMillis = System.currentTimeMillis();
+            date = new java.util.Date(currentTimeInMillis);
+        } else {
+            //Type checking the parameter
+            if (!(seconds instanceof org.python.types.Int) && !(seconds instanceof org.python.types.Float)) {
```
Comment (eliasdorneles) [inline python/common/python/time.java]: I see this is testing for float but we're not exercising it in the tests. Can you add a test passing a float as well?

Comment (prithvisingh18) [reply python/common/python/time.java]: I added another float test case too. BTW [print(time.ctime((now - (now % 3600)))](https://github.com/pybee/voc/pull/739/files#diff-3e3a3c1d82261ce04faab17d5b8024c0R121) already had a float test.

**Comment by eliasdorneles** [comment]
Thanks @prithvisingh18, merging !! :tada:
